src/Mod/PartDesign/Gui/ViewProviderTransformed.cpp

```diff
@@ -158,11 +170,171 @@ bool ViewProviderTransformed::onDelete(const std::vector<std::string> &s)
+        std::condition_variable cv;
+
+        ComputationDialog dialog(Gui::MainWindow::getInstance());
+
+        QMessageBox forceAbortBox(
+            QMessageBox::Warning,
+            QObject::tr("Operation not responding"),
+            QObject::tr("The abort operation is taking longer than expected.\nDo you want to forcibly cancel the thread? This will probably crash FreeCAD.");
+            QMessageBox::Yes | QMessageBox::No,
+            Gui::MainWindow::getInstance());
+        forceAbortBox.setDefaultButton(QMessageBox::No);
+
+        Message_ProgressRange progressRange = dialog.Start();
+        pcTransformed->setProgressRange(progressRange);
+
+        // Start computation thread
+        std::thread computeThread([&]() {
+            #if defined(__linux__) || defined(__FreeBSD__)
+            pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, nullptr);
+            pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, nullptr);
+            #endif
+            
+            try {
+                pcTransformed->recomputeFeature(true);
+            } catch (const std::exception& e) {
+                Base::Console().Error("Computation thread caught exception: %s\n", e.what());
+            } catch (...) {
+                Base::Console().Error("Computation thread caught unknown exception\n");
+            }
+
+            computationDone.store(true);
+
+            if (dialog.isVisible()) {
+                QMetaObject::invokeMethod(&dialog, "accept", Qt::QueuedConnection);
+            }
+            if (forceAbortBox.isVisible()) {
+                QMetaObject::invokeMethod(&forceAbortBox, "accept", Qt::QueuedConnection);
+            }
+
+            cv.notify_one();
+        });
+
+        {
+            // Wait for a brief moment to see if computation completes quickly
+            std::unique_lock<std::mutex> lock(mutex);  // Fixed: std::lock -> std::mutex
+            if (!cv.wait_for(lock, std::chrono::seconds(3), 
+                [&]{ return computationDone.load(); }))  // Atomic load
+            {
+                // Computation didn't finish quickly, show dialog
+                dialog.exec();
+            }
+        }
+
+        while (!computationDone.load()) {
+            // Wait for up to 3 seconds for the thread to finish
+            std::unique_lock<std::mutex> lock(mutex);
+            if (!cv.wait_for(lock, std::chrono::seconds(3),
+                [&]{ return computationDone.load(); }))
+            {
+                if (forceAbortBox.exec() == QMessageBox::Yes) {
+                    // check computationDone again just in case the thread completed while the dialog was open
+                    if (!computationDone.load()) {
+                        Base::Console().Error("Force aborting computation thread\n");
+
+                        // TODO: save the backup document now!
+
+                        #if defined(__linux__) || defined(__FreeBSD__)

```

Comment:
Have you give a look to https://dev.opencascade.org/doc/refman/html/class_o_s_d___thread.html ?

> **OSD_Thread Class Reference**
> A simple platform-intependent interface to execute and control threads.

Reply:
> It doesn't look like it gives you a way to cancel the thread

Right. 🙁